ql/src/java/org/apache/hadoop/hive/ql/optimizer/physical/NullScanTaskDispatcher.java

```diff
@@ -159,21 +165,46 @@ private void processTableScans(MapWork work, Set<TableScanOperator> tableScans)
           } else {
             managedEmptyPathMap.put(path, false);
           }
+        } else {
+          numberOfUnmanagedPaths++;
         }
       }
     }
 
+    int fetcherPoolParallelism = Math.min(MAXIMUM_ASYNC_STATUS_FETCHER_COUNT, (int)Math.sqrt(numberOfUnmanagedPaths));
+    ForkJoinPool pool = null;
+    if (fetcherPoolParallelism > 1) { // dont create thread pool for a few partitions
+      pool = new ForkJoinPool(fetcherPoolParallelism);
+    }
+    List<ForkJoinTask<Void>> tasks = new ArrayList<>(numberOfUnmanagedPaths);
     for (Entry<Path, Collection<String>> entry : candidatePathsToAliases.entrySet()) {
       Path path = entry.getKey();
       Collection<String> allowed = entry.getValue();
       Boolean isEmpty = managedEmptyPathMap.get(path);
       // if isEmpty is null, either stats are not up to date or this is external table
       if (isEmpty == null) {
-        lookupAndProcessPath(work, path, allowed);
+        if (pool != null) {
+          ForkJoinTask<Void> task = pool.submit(new Callable<Void>() {
+            @Override
+            public Void call() throws Exception {
+              lookupAndProcessPath(work, path, allowed);
+              return null;
+            }
+          });
+          tasks.add(task);
+        } else {
+          lookupAndProcessPath(work, path, allowed);
+        }
       } else {
         processPath(work, path, allowed, isEmpty);
       }
     }
+    for (ForkJoinTask<Void> t : tasks) {
```

Comment:
With regular execservice, it would be to Future::get. In case of exceptions, other futures have to be cancelled.